kratos/spatial_containers/geometrical_objects_bins.h

```diff
@@ -0,0 +1,472 @@
+        KRATOS_DEBUG_ERROR_IF(K > mNumberOfCells[2]) << "Index " << K << " is larger than number of cells in z direction : " << mNumberOfCells[2] << std::endl;
+
+        BoundingBox<Point> result;
+        
+        result.GetMinPoint()[0] = mBoundingBox.GetMinPoint()[0] + I * mCellSizes[0];
+        result.GetMinPoint()[1] = mBoundingBox.GetMinPoint()[1] + J * mCellSizes[1];
+        result.GetMinPoint()[2] = mBoundingBox.GetMinPoint()[2] + K * mCellSizes[2];
+         
+        result.GetMaxPoint()[0] = mBoundingBox.GetMinPoint()[0] + (I + 1) * mCellSizes[0];
+        result.GetMaxPoint()[1] = mBoundingBox.GetMinPoint()[1] + (J + 1) * mCellSizes[1];
+        result.GetMaxPoint()[2] = mBoundingBox.GetMinPoint()[2] + (K + 1) * mCellSizes[2];
+
+        return result;
+    }
+
+    /** This method takes a point and finds all of the objects in the given radius to it.
+     * The result contains the object and also its distance to the point.
+    */
+    template<typename TPointType>
+    void SearchInRadius(TPointType const& ThePoint, double Radius, std::vector<ResultType>& rResults) {
+        std::unordered_set<GeometricalObject*> results;
+
+        array_1d< std::size_t, 3 > min_position;
+        array_1d< std::size_t, 3 > max_position;
+
+        for(int i = 0; i < 3; i++ ) {
+            min_position[i] = CalculatePosition(ThePoint[i] - Radius, i);
+            max_position[i] = CalculatePosition(ThePoint[i] + Radius, i) + 1;
+        }
+        for(std::size_t k = min_position[2] ; k < max_position[2] ; k++){
+            for(std::size_t j = min_position[1] ; j < max_position[1] ; j++){
+                for(std::size_t i = min_position[0] ; i < max_position[0] ; i++){
+                    auto& cell = GetCell(i,j,k);
+                    SearchInRadiusInCell(cell, ThePoint, Radius, results);
+                }
+            }
+        }
+
+        rResults.clear();
+        for(auto p_object : results){
+            rResults.push_back(ResultType(p_object));
+        }
+    }
+
+    /** This method takes a point and finds the nearest object to it.
+     * If there are more than one object in the same minimum distance only one is returned
+     * Result contains a flag is the object has been found or not. 
+    */
+     template<typename TPointType>
+    ResultType SearchNearest(TPointType const& ThePoint) {
+		ResultType current_result;
+
+        array_1d< std::size_t, 3 > min_position;
+        array_1d< std::size_t, 3 > max_position;
+
+        for(int i = 0; i < 3; i++ ) {
+            min_position[i] = CalculatePosition(ThePoint[i], i);
+            max_position[i] = CalculatePosition(ThePoint[i], i) + 1;
+        }
+
+        current_result.SetDistance(std::numeric_limits<double>::max());
+
+        while (!current_result.IsObjectFound() ) {
+            for(std::size_t k = min_position[2] ; k < max_position[2] ; k++){
```

Comment:
1: Do I understand correctly that this loop will always iterate over all the cells, including the ones from the previous iteration?

2: If an object is found in one of the "corner" cells (diagonal) of the cube in the current iteration, the search will terminate. Isn't it possible that there is another object in the next layer of cells that is closer to the one already found?

Reply:
1. Yes, it loops over all of them but they should be empty. (I assume the overhead of the loop avoiding the ifs and jumps in the loop)
2. oops, I forgot to add the final search in radius... Very nice catch! Thanks!